From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Mon, 3 Apr 2023 18:54:34 +0300
Subject: [PATCH 09/10] callrec: catch exceptions from reset() when releasing
 MediaRecorder

reset() can throw an exception, despite docs not mentioning it.
---
 .../callrecord/impl/CallRecorderService.java  | 30 ++++++++++++-------
 1 file changed, 20 insertions(+), 10 deletions(-)

diff --git a/java/com/android/dialer/callrecord/impl/CallRecorderService.java b/java/com/android/dialer/callrecord/impl/CallRecorderService.java
index 3730561..1e3f91c 100644
--- a/java/com/android/dialer/callrecord/impl/CallRecorderService.java
+++ b/java/com/android/dialer/callrecord/impl/CallRecorderService.java
@@ -39,6 +39,7 @@ import com.android.dialer.callrecord.ICallRecorderService;
 import java.io.IOException;
 import java.text.SimpleDateFormat;
 import java.util.Date;
+import java.util.Objects;
 
 public class CallRecorderService extends Service {
   private static final String TAG = "CallRecorderService";
@@ -131,10 +132,9 @@ public class CallRecorderService extends Service {
       mMediaRecorder.setAudioEncoder(formatChoice == 0
           ? MediaRecorder.AudioEncoder.AMR_WB : MediaRecorder.AudioEncoder.AAC);
     } catch (IllegalStateException e) {
-      Log.w(TAG, "Error initializing media recorder", e);
-      mMediaRecorder.reset();
-      mMediaRecorder.release();
-      mMediaRecorder = null;
+      Log.e(TAG, "Error initializing media recorder", e);
+      releaseMediaRecorder();
+
       return false;
     }
 
@@ -169,9 +169,7 @@ public class CallRecorderService extends Service {
       }
     }
 
-    mMediaRecorder.reset();
-    mMediaRecorder.release();
-    mMediaRecorder = null;
+    releaseMediaRecorder();
 
     return false;
   }
@@ -182,17 +180,16 @@ public class CallRecorderService extends Service {
     if (mMediaRecorder != null) {
       try {
         mMediaRecorder.stop();
-        mMediaRecorder.reset();
-        mMediaRecorder.release();
       } catch (IllegalStateException e) {
         Log.e(TAG, "Exception closing media recorder", e);
       }
 
+      releaseMediaRecorder();
+
       Uri uri = ContentUris.withAppendedId(
           MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, mCurrentRecording.mediaId);
       getContentResolver().update(uri, CallRecording.generateCompletedValues(), null, null);
 
-      mMediaRecorder = null;
       mCurrentRecording = null;
     }
     return recording;
@@ -219,4 +216,17 @@ public class CallRecorderService extends Service {
   public static boolean isEnabled(Context context) {
     return context.getResources().getBoolean(R.bool.call_recording_enabled);
   }
+
+  private void releaseMediaRecorder() {
+    Objects.requireNonNull(mMediaRecorder);
+    try {
+      mMediaRecorder.reset();
+    } catch (Exception e) {
+      Log.e(TAG, "unable to reset media recorder", e);
+    }
+
+    mMediaRecorder.release();
+
+    mMediaRecorder = null;
+  }
 }
